// mr-state/state/selectors.ts
import { Name, Package, Path, RootPackageJson, SubPackageJson } from 'mr-types';
import { basename, dirname } from 'path';
import * as _ from 'ramda';
import { createSelector, Selector } from 'reselect';
import { Model } from './types';

export const getRootPackage = (state: Model): Path => state.rootPackage;
export const getRootPackageJson: Selector<Model, RootPackageJson> = createSelector(
  getRootPackage,
  require,
);

export const getSubPackages = (state: Model): Path[] => state.subPackages;
export const getSubPackageJsons: Selector<Model, SubPackageJson[]> = createSelector(
  getSubPackages,
  _.map(require),
);

export const getFocus = (state: Model): Name => state.focus;
export const getSubPackageNames: Selector<Model, string[]> =
  createSelector(
    getSubPackageJsons,
    _.map(_.prop('name')),
  );

export const getSubPackageVersions =
  createSelector(
    getSubPackageJsons,
    (subs) =>
      _.mapObjIndexed(
        _.propOr('0.1.0', 'version'),
        _.indexBy(_.propOr('', 'name'), subs),
      ),
  );

export const getPackageMapObjects = createSelector(
  getRootPackageJson,
  getSubPackages,
  getSubPackageVersions,
  (rootJson, subs, depVersions) => {
    return _.map((packagePath) => {
      const packageJson = require(packagePath);
      const pickDeps = _.pick(packageJson.dependencies || []);
      return {
        name: packageJson.name,
        version: packageJson.version,
        cwd: dirname(packagePath),
        scripts: {
          ...rootJson.scripts,
          ...packageJson.scripts,
        },
        dependencies: {
          ...pickDeps(depVersions),
          ...pickDeps(rootJson.dependencies),
        },
        ..._.pick(
          ['bugs', 'repository', 'author', 'license', 'homepage'],
          rootJson,
        ),
      };
    }, subs);
  },
);

export const getPackageMap = createSelector(
  getPackageMapObjects,
  _.indexBy(_.prop('name')),
);

export const getFocusedPackageJson: Selector<Model, SubPackageJson | undefined> =
  createSelector(
    getFocus,
    getSubPackageJsons,
    (name, subs) => _.find(_.whereEq({ name }), subs),
  );

export const getFocusedPackagePath: Selector<Model, Path | undefined> =
  createSelector(
    getFocus,
    getPackageMap,
    (name, obj) => (obj[name] || {}).cwd,
  );

export const getFocusedPackage: Selector<Model, Package | undefined> =
  createSelector(
    getFocus,
    getPackageMap,
    (name, obj) => (obj[name] || {}),
  );

export const getSubPackageDeps =
  createSelector(
    getSubPackageJsons,
    (subs) => _.mapObjIndexed(
      _.propOr([], 'dependencies'),
      _.indexBy(_.propOr('', 'name'), subs),
    ),
  );
