// mr-state/state/types.ts
import { Name, Path } from 'mr-types';
import { Action } from 'redux';
// =============================================================================
// Model
//
// Define the shapes our model.
//

export interface Model {
  rootPackage: Path;
  subPackages: Path[];
  focus: Name;
}

// =============================================================================
// Types
//
// Types identify the redux actions.
//
export enum Type {
  // Set the root of the current mono repo
  SET_ROOT = 'SET_ROOT',
  // Set the Sub packages associated with the root package
  SET_SUBS = 'SET_SUBS',
  // Set the name of the package to focus on.
  SET_FOCUS = 'SET_FOCUS',
  // Actions that we're not concerned with in our reducer
  UNDEFINED = 'UNDEFINED',
}

// =============================================================================
// Actions
//
// Define the shapes our actions.
//

export interface EmptyAction<T extends Type> {
  type: T;
}

export interface PayloadAction<T extends Type, P> {
  type: T;
  payload: P;
}
