// mr-config/index.ts
import rc from 'rc';

const defaults = {
  rootPackagePath: false,
  rootPackageJson: false,
  subPackagePaths: false,
  initState: false,
  root: process.cwd(),
  config: false,
  verbose: false,
  help: false,
};

export const config = rc('mrtools', defaults);

if (config.verbose) {
  // tslint:disable-next-line
  console.log(JSON.stringify(config, null, 2));
}

if (config.help || config.h) {
  // tslint:disable-next-line
  console.log(`
  mr-tools [OPTION] <root-directory>

  Options
           --fe               start the graphical frontend

           --rootPackagePath  show the path to the root package.json
           --subPackagePaths  show the paths to the sub packages
           --initState        show the initial state and exit
           --subs             show the sub package.json files before interpolation
           --subsAfter        show the sub package.json files after interpolation
           --names            show the sub package names

    -v     --verbose          show debugging output
    -h     --help             show this help

  Source: https://github.com/MrRacoon/mr-tools
  `);

  process.exit(0);
}
