// mr-dashboard/index.ts
import blessed from 'blessed';
import contrib from 'blessed-contrib';
import { config } from 'mr-config';
import { actions, Actions, Model, select } from 'mr-state';
import { Name, PackageJson } from 'mr-types';
import _ from 'ramda';
import { Dispatch, Store } from 'redux';

const debug = config.verbose;

export default function(store: Store<Model>) {
  const screen = blessed.screen();

  const grid = new contrib.grid({
    screen,
    rows: 12,
    cols: 12,
  });

  const logger = grid.set(8, 0, 4, 2, contrib.log, {
    label: 'Master Logger',
    fg: 'cyan',
  });

  const name = grid.set(0, 0, 8, 2, contrib.tree, {
    label: 'Package Names',
    fg: 'green',
    vi: true,
    style: {
      fg: 'white',
      selected: {
        bg: 'yellow',
        fg: 'black',
      },
    },
    template: {
      lines: true,
    },
    mouse: true,
    keys: [ 'space', 'enter', 'l', '+' ],
  });

  const focussed = grid.set(0, 2, 12, 4, contrib.tree, {
    label: 'Focussed Package',
    style: {
      fg: 'yellow',
      selected: {
        bg: false,
        fg: 'yellow',
      },
    },
    vi: true,
    keys: [],
    template: {
      extend: ' ',
      retract: ' ',
      lines: true,
    },
  });

  const data = select.getSubPackageNames(store.getState())
    .reduce((obj, pkgName: string) => ({
      ...obj,
      [pkgName]: {},
    }), {});

  name.setData({
    extended: true,
    children: data,
  });

  interface Node {
    name: string;
  }

  // ===========================================================================
  // Focussed Package

  function expandJson(json: object): object {
    return _.mapObjIndexed((val, key) => {
      if (Array.isArray(val)) {
        if (debug) { logger.log(`${key} is array`); }
        return {
          extended: true,
          children: _.reduce((obj, v) => {
            return {
              ...obj,
              [v]: {},
            };
          } , {}, val),
        };
      }
      if (typeof val === 'object') {
        if (debug) { logger.log(`${key} is object`); }
        return {
          extended: true,
          children: expandJson(val),
        };
      }
      if (debug) { logger.log(`${key} is val ${val}`); }
      return {
        extended: true,
        children: {
          [val]: {},
        },
      };
    }, json);
  }

  function setFocus(state: Model) {
    const focussedPackage = select.getFocusedPackageJson(state);
    if (focussedPackage) {
      focussed.setData({
        extended: true,
        children: expandJson(focussedPackage),
      });
    }
  }

  name.on('select', (node: Node) => {
    if (debug) { logger.log(`Selected: ${node.name}`); }
    store.dispatch(actions.setFocus(node.name));
    setFocus(store.getState());
    screen.render();
  });

  // ===========================================================================
  // KeyBindings
  screen.key([
    'escape', 'q', 'C-c',
  ], (ch, key) => {
    switch (ch) {
      default:
        return process.exit(0);
    }
  });

  // ===========================================================================
  // Last Minute adjustments

  // Set initial focus to be the package list.
  name.focus();

  // Drop a message in the master logger.
  logger.log('rendering');

  // Call render.
  screen.render();

  // Return pointers to each view.
  return { screen, grid, focussed, name };
}
