// mr-dashboard/index.ts
import blessed from 'blessed';
import contrib from 'blessed-contrib';
import { Model, select } from 'mr-state';
import { Store } from 'redux';

export default function(store: Store<Model>) {
  const screen = blessed.screen();

  const grid = new contrib.grid({
    screen,
    rows: 12,
    cols: 12,
  });

  const logger = grid.set(8, 0, 4, 4, contrib.log, {
    label: 'Master Logger',
    fg: 'cyan',
  });

  const name = grid.set(0, 0, 8, 4, contrib.tree, {
    label: 'Package Names',
    fg: 'green',
    vi: true,
  });

  const data = select.getSubPackageNames(store.getState())
    .reduce((obj, pkgName: string) => ({
      ...obj,
      [pkgName]: {},
    }), {});

  name.setData({
    extended: true,
    children: data,
  });

  interface Node {
    name: string;
  }

  name.on('select', (node: Node) => {
    logger.log(node.name);
  });

  name.focus();

  screen.key(['escape', 'q', 'C-c'], (ch, key) => {
    return process.exit(0);
  });

  logger.log('rendering');

  screen.render();

  return {
    screen, grid,
    name,
  };
}
